File: server/routes/competitors.js
const express = require('express');
const router = express.Router();
const { validateSiteParams, asyncHandler } = require('../middleware/validate');
const { success, error } = require('../utils/response');
const { competitorAnalysis, opportunityGaps, competitorContext } = require('../services/competitorService');

// ── Validate business_type query param ──────────────────────────────────────
const VALID_TYPES = ['restaurant','pharmacy','grocery','clinic','hospital','education','fitness','bank','hotel','tea'];

function validateBusinessType(req, res, next) {
  const type = req.query.business_type || 'restaurant';
  if (!VALID_TYPES.includes(type)) {
    return res.status(400).json(error(`Invalid business_type. Use one of: ${VALID_TYPES.join(', ')}`));
  }
  req.businessType = type;
  next();
}

// GET /api/competitors?lng=&lat=&radius=&business_type=
router.get('/', validateSiteParams, validateBusinessType, asyncHandler(async (req, res) => {
  const data = await competitorAnalysis({
    ...req.siteParams,
    business_type: req.businessType,
  });
  res.json(success(data));
}));

// GET /api/competitors/gaps?lng=&lat=&radius=&business_type=
router.get('/gaps', validateSiteParams, validateBusinessType, asyncHandler(async (req, res) => {
  const data = await opportunityGaps({
    ...req.siteParams,
    business_type: req.businessType,
  });
  res.json(success(data));
}));

// GET /api/competitors/context?comp_lng=&comp_lat=&site_lng=&site_lat=
router.get('/context', asyncHandler(async (req, res) => {
  const { comp_lng, comp_lat, site_lng, site_lat } = req.query;
  const parsed = [comp_lng, comp_lat, site_lng, site_lat].map(parseFloat);
  if (parsed.some(isNaN)) {
    return res.status(400).json(error('comp_lng, comp_lat, site_lng, site_lat are all required numeric values'));
  }
  const data = await competitorContext({
    comp_lng: parsed[0], comp_lat: parsed[1],
    site_lng:  parsed[2], site_lat:  parsed[3],
  });
  res.json(success(data));
}));

module.exports = router;

File: backend-py/app/routes/competitors.py
"""Port of server/routes/competitors.js — all endpoints require auth."""
from fastapi import APIRouter, Depends

from ..auth import require_auth
from ..common import ApiError, success, validate_site_params
from ..services.competitor import (
    VALID_TYPES, competitor_analysis, competitor_context, opportunity_gaps,
)

router = APIRouter(prefix="/api/competitors", dependencies=[Depends(require_auth)])


def _business_type(business_type):
    bt = business_type or "restaurant"
    if bt not in VALID_TYPES:
        raise ApiError(f"Invalid business_type. Use one of: {', '.join(VALID_TYPES)}", 400)
    return bt


@router.get("")
def competitors(lat: str = None, lng: str = None, radius: str = None, business_type: str = "restaurant"):
    params = validate_site_params(lat, lng, radius)
    params["business_type"] = _business_type(business_type)
    return success(competitor_analysis(params))


@router.get("/gaps")
def gaps(lat: str = None, lng: str = None, radius: str = None, business_type: str = "restaurant"):
    params = validate_site_params(lat, lng, radius)
    params["business_type"] = _business_type(business_type)
    return success(opportunity_gaps(params))


@router.get("/context")
def context(comp_lng: str = None, comp_lat: str = None, site_lng: str = None, site_lat: str = None):
    try:
        vals = [float(comp_lng), float(comp_lat), float(site_lng), float(site_lat)]
    except (TypeError, ValueError):
        raise ApiError("comp_lng, comp_lat, site_lng, site_lat are all required numeric values", 400)
    data = competitor_context(comp_lng=vals[0], comp_lat=vals[1], site_lng=vals[2], site_lat=vals[3])
    return success(data)
